### ISSUE TITLE
Flash usage of sequential storage

### ISSUE THREAD

**Issue by HaoboGu** [tweedegolf/sequential-storage#34]
Hi there, I'm using sequential-storage for more usages in my project, I found that `sequential-storage` uses a lot of flash space:

<img width="910" alt="image" src="https://github.com/tweedegolf/sequential-storage/assets/8640918/e35a6e71-933c-4188-a72c-80baff2969f0">

I implemented `StorageItem` for just 4 structs. Is there any way to reduce flash usage? thanks!

**Comment by HaoboGu** [comment]
yeah. 

`store_item_inner`s are still there.

<img width="908" alt="image" src="https://github.com/tweedegolf/sequential-storage/assets/8640918/a53a6103-f764-4946-9ab8-3b9fcc98ffb9">

my cargo profile config:

```toml
[profile.release]
codegen-units = 1       # better optimizations
debug = true            # no overhead for bare-metal
opt-level = "z"         # optimize for binary size
overflow-checks = false
lto = false
```

and rust flags in .cargo/config:

```toml
rustflags = [
  "-C", "inline-threshold=0"
]
```

**Comment by diondokter** [comment]
Thanks for checking! Strange how the numbers don't change much...

I'll try to look into it when I have time

**Comment by diondokter** [comment]
I've made my own repo for checking out the size here: https://github.com/diondokter/s-s-size

**Comment by diondokter** [comment]
Ok, so after some investigation one thing that causes extra flash usage is generics and how I did the caching.

For ease of use I implemented the relevant cache traits for `impl<T: Cache> Cache for &mut T` as well. This is nice from a usage point, but it makes it very easy to have multiple cache impls in the system which all monomorphise to their own functions and futures.

This means that it's possible to see the same function three times with:
- `NoCache`
- `&mut NoCache`
- `&mut &mut NoCache`

If I force everyone to use `&mut impl Cache`, then that saves ~2-3kb or 25% of the total binary size of my testing repo.
This enforcement is a breaking change though...

If you use multiple types as the Item type, then the same thing happens.
Maybe that can be fixed if I change the trait for it, but that's also a breaking change.

**Comment by diondokter** [comment]
Using a little more what?

**Comment by HaoboGu** [comment]
> Is it possible to optimize flash using a little more?

-> Is it possible to optimize flash usage a little more?


ah sorry for the typo

**Comment by diondokter** [comment]
Ah right.

Well... maybe? I'm looking into it.
The multiple cache impls was most obvious.
I can try make an optimization so that using multiple item types doesn't cost a lot.

Other than that will be difficult. Debugging futures is hard. It's difficult to see the statemachines that the compiler produces and the generated assembly is even harder to read than normal non-async assembly.

**Comment by HaoboGu** [comment]
Yeah, I agree that debugging async rust is quite hard. And I'm looking forward to the further improvement, thanks for your great work ❤️

**Comment by diondokter** [comment]
@HaoboGu I think I've got all the size optimizations in the PR that I want to make.
To do more I had to change the map API too, but the new API is probably closer to what people expect when they use the crate, so that's a good thing.

Could you try it out?

**Comment by HaoboGu** [comment]
Sure, will try soon!

**Comment by HaoboGu** [comment]
Just tried it out. The api is more clear to me than before, and binary size of s-s was reduced by ~3KB, from ~11KB -> ~8KB in my project. That's great!

Thank you @diondokter very much for your work, I really appreciate it.

**Comment by diondokter** [comment]
Thanks for the feedback!

There's one other issue I want to work on before I release the next version. But for the mean time you can use the master branch
